CLE-3128: Django42 upgrade fixes

DelboyJay opened the pull request:
Changes that are required for projects that consume this one to run Django 4.2.

### Added
- Support for Python 3.9 to 3.13
- Support for Django 3.2, 4.2, 5.2
- changelog to track future changes.
- Pre-commit checks and run them as part of the Github actions workflow.
- Pre-commit checks for missing migrations
- Make file contains command `db-setup` to test running all migrations on an empty database

### Changed
- Use Ruff as linter
 
### Removed
- Dropped support for Django 2.2 and python 3.7 & 3.8

DelboyJay commented on .github/actions/drf-integrations-framework-build/action.yml:
```diff
@@ -7,10 +7,10 @@ runs:
   steps:
   - name: Check out repository
-    uses: actions/checkout@v3
+    uses: actions/checkout@v4
```
fixes a warning from github about the actions being forced to use node.js 20.

DelboyJay commented on example/settings.py:
```diff
@@ -19,7 +19,7 @@
 # See https://docs.djangoproject.com/en/3.0/howto/deployment/checklist/
 
-SECRET_KEY = "mysecret"
+SECRET_KEY = "231cc7b8-63ce-4a87-98ea-4c663ee9557d:d7bcde56-e812-4ef0-aa07-3b22f3066d2f"
```
This is just for an example app so does not matter but fixed a django check warning I was seeing.

DelboyJay commented on tests/conftest.py:
```diff
@@ -9,10 +9,10 @@
-    TestInternalWithFormIntegration,
-    TestLocalIntegration,
-    TestLocalWithFormIntegration,
+    InternalIntegrationTest,
```
having test at the start of the name confuses pytest as these are not true pytest classes

lewis-kori commented on drf_integrations/migrations/0006_alter_grant_redirect_uri.py:
```diff
@@ -0,0 +1,27 @@
+        ("drf_integrations", "0005_alter_applicationinstallation_id"),
+    ]
+
+    # Django does not explicitly follow the dependency definitions under the
+    # `dependencies` variables as you might expect.
+    # It is imperative that these migrations are run BEFORE
+    # oauth2_provider.0004_auto_20200902_2022, or we get errors because the
+    # drf_integrations.Application model cannot be found
```
this should be documented as part of the setup instructions.

DelboyJay replied:
I've added a section under README.md to mention this when updating the migrations in case we come across it in the future.

lewis-kori commented on CHANGELOG.md:
Can we rely on GitHub releases to generate the changelogs? One can develop that there as you make a release. It'd be an ideal setup as one can see the change log through the GitHub interface as you track the packages.

DelboyJay replied:
Technically we should do both. CHANGELOG.md is a dev technical level and github releases is for user-based top-level overview.

lewis-kori reviewed (approved):
Looks good. I think we should add a `CONTRIBUTING.rst` as a guideline for how open source contributors can get involved with the project.

example: https://github.com/saltpay/drf-jwt-auth/blob/master/CONTRIBUTING.rst

lewis-kori commented on README.md:
```diff
@@ -10,10 +10,10 @@ Framework will probably simplify that for you. DRF Integrations Framework will h
-- Django 2.2+
-- Django REST Framework 3.9.2+
-- Django OAuth Toolkit 1.3-1.4
+- Python >=3.9+, <3.13
```
if the minimum version is 3.9 then the github action runner for 3.7 should be removed. That explains why it's stuck.

DelboyJay committed:
- minimal changes to support django 4.2 and Python 3.12

DelboyJay committed:
- GitHub to use postgres 17
- pin setuptools to <81 as recommended by warning messages from tox
- remove USE_L10N because it will be deprecated.
- fixed GitHub tests actions so that we install the poetry environment for tox and all dependencies are in one place instead of being duplicated under tox.ini in the `deps` section
-

DelboyJay committed:
- change poetry buildsystem section to match Platform's

DelboyJay committed:
- set version to `0.9.0a1`

DelboyJay committed:
- move to ruff linting

DelboyJay committed:
- run ruff linting on all files and make corrections

DelboyJay committed:
- updated the makefile with some more command to standardize across repos.

DelboyJay committed:
- Added python3.13 tests to tox

DelboyJay committed:
- Fix more linting errors

DelboyJay committed:
- Changed github to run tox tests for python 3.9, 3.10, and 3.11

DelboyJay committed:
- added Django 5.2 tests to tox

DelboyJay committed:
- updated the CHANGELOG.md

lewis-kori commented on .github/workflows/release.yml:
```diff
@@ -2,16 +2,39 @@ name: Release
   release:
-    types:
-      - created
+    types: [created]
```
Do we have a version of the package in the [testing index](https://test.pypi.org/)? If so, we should have a r[elease, preleased workflow](https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows#release) for the different targets.

DelboyJay replied:
No I'm not aware of that.

lewis-kori commented on .pre-commit-config.yaml:
```diff
@@ -1,41 +1,43 @@
 repos:
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.12.12
```
```suggestion
    rev: v0.13.0
```

lewis-kori commented on pyproject.toml:
```diff
@@ -5,7 +5,7 @@
 [tool.poetry]
 name = "drf-integrations-framework"
-version = "0.8.6"
+version = "0.9.0a1"
```
Are we using [semantic versioning](https://semver.org/) or [PEP440](https://blog.inedo.com/python/best-practices-for-versioning-python-packages-in-the-enterprise/)?

DelboyJay replied:
0.9.0a1 is PEP440 and I'm only using a1 for now because I want to do proper testing on staging with the Django 4.2 branch first before I release properly as 0.9.0

lewis-kori commented on setup.py:
```diff
@@ -1,13 +1,14 @@
+
 CURRENT_PYTHON = sys.version_info[:2]
-REQUIRED_PYTHON = (3, 7)
+REQUIRED_PYTHON = (3, 12)
```
I've seen in the CHANEGELOG that we support python 3.13. will this break that or is it a minimal requirement to build the package?

DelboyJay replied:
Well spotted Lewis, thanks for this. I've deleted that setup.py file as we don't need to worry about this anymore as tox is testing all python versions up to and including 3.13.

lewis-kori reviewed (approved):
Nicely done!

DelboyJay committed:
Update .pre-commit-config.yaml

Co-authored-by: Lewis Kori <39276332+lewis-kori@users.noreply.github.com>

DelboyJay committed:
- updated the CHANGELOG.md
- setup.cfg is no longer needed as this is covered by ruff now.
- setup.py is not needed as we support upto python 3.13
- updated classifiers in pyproject.toml
